### PR TITLE
Added sampling support for opentracing

### PR DESCRIPTION
**Pull request by vikramraman**

### CODE REVIEW

**Comment by vikramraman** [comment]
Note: Build is failing as it relies on an unreleased version of `wavefront-sdk-java`.

File: src/main/java/com/wavefront/opentracing/WavefrontTracer.java [hunk 1]
```diff
@@ -84,6 +87,26 @@ public SpanBuilder buildSpan(String operationName) {
...
+  boolean sample(String operationName, long traceId, long duration) {
+    if (samplers == null || samplers.isEmpty()) {
+      return true;
+    }
+    boolean earlySampling = (duration == 0);
+    for (Sampler sampler : samplers) {
+      boolean doSample = earlySampling == sampler.isEarly();
+      if (doSample && sampler.sample(operationName, traceId, duration)) {
+        if (logger.isLoggable(Level.FINER)) {
+          logger.finer(sampler.getClass().getSimpleName() + "=" + true + "op=" + operationName);
```
Comment (basilisk487) [inline src/main/java/com/wavefront/opentracing/WavefrontTracer.java]: nit-pick: space before op= ?

File: src/main/java/com/wavefront/opentracing/WavefrontTracer.java [hunk 2]
```diff
@@ -84,6 +87,26 @@ public SpanBuilder buildSpan(String operationName) {
...
+        }
+        return true;
+      }
+      if (logger.isLoggable(Level.FINER)) {
+        logger.finer(sampler.getClass().getSimpleName() + "=" + false + "op=" + operationName);
```
Comment (basilisk487) [inline src/main/java/com/wavefront/opentracing/WavefrontTracer.java]: ditto

Comment (basilisk487) [reply src/main/java/com/wavefront/opentracing/WavefrontTracer.java]: lgtm otherwise

File: pom.xml [hunk 3]
```diff
@@ -90,7 +90,7 @@
         <dependency>
             <groupId>com.wavefront</groupId>
             <artifactId>wavefront-sdk-java</artifactId>
-            <version>1.0</version>
+            <version>1.1-SNAPSHOT</version>
```
Comment (sushantdewan123) [inline pom.xml]: let's change this to 1.1

File: src/main/java/com/wavefront/opentracing/WavefrontSpan.java [hunk 4]
```diff
@@ -70,6 +79,18 @@ public WavefrontSpan setTag(String key, Number value) {
   private synchronized WavefrontSpan setTagObject(String key, Object value) {
     if (key != null && !key.isEmpty() && value != null) {
       tags.add(Pair.of(key, value.toString()));
+
+      if (Tags.SAMPLING_PRIORITY.getKey().equals(key) && value instanceof Number) {
```
Comment (sushantdewan123) [inline src/main/java/com/wavefront/opentracing/WavefrontSpan.java]: Add one line comment for what this code is doing/overridding

File: src/main/java/com/wavefront/opentracing/WavefrontSpan.java [hunk 5]
```diff
@@ -70,6 +79,18 @@ public WavefrontSpan setTag(String key, Number value) {
...
+        int priority = ((Number) value).intValue();
+        forceSampling = priority > 0 ? Boolean.TRUE : Boolean.FALSE;
+        spanContext = spanContext.withSamplingDecision(forceSampling);
+      }
+      if (forceSampling == null && Tags.ERROR.getKey().equals(key)) {
```
Comment (sushantdewan123) [inline src/main/java/com/wavefront/opentracing/WavefrontSpan.java]: Add one line comment for what this code is doing/overridding

File: src/main/java/com/wavefront/opentracing/WavefrontSpanBuilder.java [hunk 6]
```diff
@@ -131,24 +131,32 @@ public Span start() {
...
     UUID spanId = UUID.randomUUID();
-    return new WavefrontSpanContext(traceId, spanId);
+    WavefrontSpanContext traceCtx = traceAncestry();
+    UUID traceId = traceCtx == null ? UUID.randomUUID() : traceCtx.getTraceId();
+    Boolean sampling = traceCtx == null ? null : traceCtx.getSamplingDecision();
+    return new WavefrontSpanContext(traceId, spanId, null, sampling);
   }
 
-  private UUID traceAncestry() {
+  private WavefrontSpanContext traceAncestry() {
```
Comment (sushantdewan123) [inline src/main/java/com/wavefront/opentracing/WavefrontSpanBuilder.java]: should be marked @Nullable ?

File: src/main/java/com/wavefront/opentracing/WavefrontSpanBuilder.java [hunk 7]
```diff
@@ -131,24 +131,32 @@ public Span start() {
...
+    return new WavefrontSpan(tracer, operationName, ctx, startTimeMicros, startTimeNanos, parents,
+        follows, tags);
   }
 
   private WavefrontSpanContext createSpanContext() {
-    UUID traceId = traceAncestry();
     UUID spanId = UUID.randomUUID();
-    return new WavefrontSpanContext(traceId, spanId);
+    WavefrontSpanContext traceCtx = traceAncestry();
+    UUID traceId = traceCtx == null ? UUID.randomUUID() : traceCtx.getTraceId();
```
Comment (sushantdewan123) [inline src/main/java/com/wavefront/opentracing/WavefrontSpanBuilder.java]: Usually try to add paranthesis for easier readability
see below
UUID traceId = (traceCtx == null) ? UUID.randomUUID() : traceCtx.getTraceId();

File: src/main/java/com/wavefront/opentracing/WavefrontSpanBuilder.java [hunk 8]
```diff
@@ -131,24 +131,32 @@ public Span start() {
...
+    Boolean sampling = traceCtx == null ? null : traceCtx.getSamplingDecision();
```
Comment (sushantdewan123) [inline src/main/java/com/wavefront/opentracing/WavefrontSpanBuilder.java]: same comment
Boolean sampling = (traceCtx == null) ? null : traceCtx.getSamplingDecision();

File: src/test/java/com/wavefront/opentracing/WavefrontSpanBuilderTest.java [hunk 9]
```diff
@@ -55,4 +60,118 @@ public void testMultiValuedTags() {
...
+    WavefrontSpanContext parentCtx = new WavefrontSpanContext(UUID.randomUUID(), UUID.randomUUID(),
+        null, Boolean.TRUE);
+
+    // Verify span created asChildOf parentCtx inherits parent sampling decision
+    WavefrontSpan span = (WavefrontSpan) tracer.buildSpan("testOp").
+        asChildOf(parentCtx).
+        start();
+
+    assertFalse(tracer.sample(span.getOperationName(), span.context().getTraceId()
+        .getLeastSignificantBits(), 0));
```
Comment (sushantdewan123) [inline src/test/java/com/wavefront/opentracing/WavefrontSpanBuilderTest.java]: nit: line break after .

File: src/test/java/com/wavefront/opentracing/WavefrontSpanBuilderTest.java [hunk 10]
```diff
@@ -55,4 +60,118 @@ public void testMultiValuedTags() {
...
+    WavefrontSpanContext parentCtx = new WavefrontSpanContext(UUID.randomUUID(), UUID.randomUUID(),
+        null, Boolean.FALSE);
+
+    // Verify span created asChildOf parentCtx inherits parent sampling decision
+    WavefrontSpan span = (WavefrontSpan) tracer.buildSpan("testOp").
+        asChildOf(parentCtx).
+        start();
+
+    assertTrue(tracer.sample(span.getOperationName(), span.context().getTraceId()
+        .getLeastSignificantBits(), 0));
```
Comment (sushantdewan123) [inline src/test/java/com/wavefront/opentracing/WavefrontSpanBuilderTest.java]: nit: line break after .
do it else where as well.

**Pull request merged**